idx/memory/tag_query.go

```diff
@@ -171,247 +248,751 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+				continue
+			}
+
+			key := tag[:equal]
+			// this tag has already been pushed into tagCh, so we can stop evaluating
+			if _, ok := resultsCache[key]; ok {
+				continue
+			}
+
+			if q.filterTagBy == PREFIX_TAG {
+				if !strings.HasPrefix(key, q.tagPrefix) {
+					continue
+				}
+			} else if q.filterTagBy == MATCH_TAG {
+				if _, ok := q.tagMatch.missCache.Load(key); ok || !q.tagMatch.value.MatchString(tag) {
+					if !ok {
+						q.tagMatch.missCache.Store(key, struct{}{})
+					}
+					continue
+				}
+			}
+			metricTags[key] = struct{}{}
+		}
+
+		if matchName {
+			if _, ok := resultsCache["name"]; !ok {
+				metricTags["name"] = struct{}{}
+			}
+		}
+
+		// if some tags satisfy the current tag filter condition then we run
+		// the metric through all tag expression tests in order to decide
+		// whether those tags should be part of the final result set
+		if len(metricTags) > 0 {
+			if q.testByAllExpressions(id, def, matchName) {
+				for key := range metricTags {
+					select {
+					case tagCh <- key:
+					case <-stopCh:
+						// if execution of query has stopped because the max tag
+						// count has been reached then tagCh <- might block
+						// because that channel will not be consumed anymore. in
+						// that case the stop channel will have been closed so
+						// we so we exit here
+						break IDS
+					}
+					resultsCache[key] = struct{}{}
+				}
+			} else {
+				// check if we need to stop
+				select {
+				case <-stopCh:
+					break IDS
+				default:
+				}
+			}
+		}
+	}
+
+	q.wg.Done()
+}
+
+// determines whether the given tag prefix/tag match will match the special
+// tag "name". if it does, then we can omit some filtering because we know
+// that every metric has a name
+func (q *TagQuery) tagFilterMatchesName() bool {
+	matchName := false
+
+	if q.filterTagBy == PREFIX_TAG || q.startWith == PREFIX_TAG {
+		if strings.HasPrefix("name", q.tagPrefix) {
+			matchName = true
+		}
+	} else if q.filterTagBy == MATCH_TAG || q.startWith == MATCH_TAG {
+		if q.tagMatch.value.MatchString("name") {
+			matchName = true
+		}
+	} else {
+		// some tag queries might have no prefix specified yet, in this case
+		// we do not need to filter by the name
+		matchName = true
+	}
+
+	return matchName
+}
+
+// RunGetTags executes the tag query and returns all the tags of the
+// resulting metrics
+func (q *TagQuery) RunGetTags(index TagIndex, byId map[string]*idx.Archive) map[string]struct{} {
+	q.index = index
+	q.byId = byId
+
+	maxTagCount := int32(math.MaxInt32)
+
+	// start a thread to calculate the maximum possible number of tags.
+	// this might not always complete before the query execution, but in most
+	// cases it likely will. when it does end before the execution of the query,
+	// the value of maxTagCount will be used to abort the query execution once
+	// the max number of possible tags has been reached
+	q.wg.Add(1)
+	go atomic.StoreInt32(&maxTagCount, int32(q.getMaxTagCount()))
```

Comment:
it's not really clear why it's advantageous to do this. It seems executing getMaxTagCount is almost like just running the entire query, i must be missing something. can you clarify in the code please, thanks

Reply:
If there's a case where we have a large index with low cardinality and an expression query comes in that results in a large initial result set, then we do not need to evaluate the whole query if we know in advance how many tags the result set could possibly contain at max.

Example:
We have tags `tagaaa; tagbbb; tagccc` and a million series that have some or all of those tags with various values and the expression query that's coming in results in the full million series. If we'd want to be sure that we have all the tags associated with the results of that expression query we'd first need to evaluate each one of the metrics (which are all going to pass the tests), then extract the tags and store them in a map, only to have those 3 tags in the end. We need to evaluate all of them because we can't know if there's a 4th tag somewhere in-there, so we need to check each one. 
If right at the beginning of the query we kick of this other go-routine it can just take a short look at the index and conclude that the result set can't be larger than 3, because there are only 3 tags in the whole index, so once those 3 tags have been collected via the expression evaluation no more IDs need to be evaluated and we can abort the expression evaluation. Which can result in a huge short cut under such conditions.